React Revamp: Add keys in React lesson (#25669)
TheOdinProject/curriculum: react/getting_started_with_react/keys_in_react.md

### Introduction

In this lesson we will cover keys in React. Keys are special props for our components and we'll learn why they are used.

> ### Lesson Overview

Reviewer: ```suggestion
### Lesson overview
```

Styleguide now uses sentence case

By the end of the lesson you should be able to answer the following:

> - Why are keys and why does React need them. 

Reviewer: ```suggestion
- What are keys and why does React need them?
```

turned into a question

> - What are good and bad examples of keys. 

Reviewer: ```suggestion
- What are good and bad examples of keys?
```

also turned into a question. Alternatively it could be written as a sentence:
> What good and bad examples of keys are

> ### Why Does React Need Keys?

Reviewer: ```suggestion
### Why does React need keys?
```

Style guide now uses sentence case

> In the upcoming lessons as you learn more about the internal workings of React, more specifically the rerendering process, you will understand the importance of keys. For now we will keep it simple. Recall the previous lesson on rendering lists. We used the `.map()` method to iterate over an array of data and return a list of elements. Now imagine, if any of the item in the list were to change, how would React know which item to update?

Reviewer: ```suggestion
In the upcoming lessons as you learn more about the internal workings of React, more specifically the rerendering process, you will understand the importance of keys. For now we will keep it simple. 

In the previous lesson on rendering lists we used the `.map()` method to iterate over an array of data and return a list of elements. Now imagine, if any of the item in the list were to change, how would React know which item to update?
```

Do you think this sounds better? I felt like this needed to be split out more and I changed it a bit because the sentence felt a bit chopped up. What do you think?

Author: I like this.

> In the upcoming lessons, as you learn more about the internal workings of React, more specifically, the rerendering process, you will understand the importance of keys. For now, we will keep it simple. Recall the previous lesson on rendering lists. We used the `.map()` method to iterate over an array of data and return a list of elements. Now imagine, if any of the item in the list were to change, how would React know which item to update?

Reviewer: ```suggestion
In the upcoming lessons as you learn more about the internal workings of React, more specifically the rerendering process, you will understand the importance of keys. For now we will keep it simple. Recall the previous lesson on rendering lists. We used the `.map()` method to iterate over an array of data and return a list of elements. Now imagine, if any of the item in the list were to change, how would React know which item to update?
```

nit, there are a lot of commas there

> If the list where to change, one of two things _should_ happen:

Reviewer: ```suggestion
If the list were to change, one of two things _should_ happen:
```

> 1. either, we completely re-render the entire list,

Reviewer: ```suggestion
1. we completely re-render the entire list, or:
```

> 2. or, hunt down that specific items that were changed and only re-render those.

Reviewer: ```suggestion
2. we hunt down that specific items that were changed and only re-render those.
```

Assuming we want to hunt down that one specific item that was changed and NOT re-render the entire list. We need something to track that specific item. We can track down a specific item by using a `key`.

> When the list is updated for whatever reason, (either from a server or a user interaction), React matches the `key`s of each of the previous list to the updated list. If there were any changes, React will only update the items that have changed.

Reviewer: ```suggestion
When the list is updated for whatever reason, (either from a server or a user interaction), React matches the `keys` of each of the previous list to the updated list. If there were any changes, React will only update the items that have changed.
```

since we use `keys` in the next sentence as well

As long as `keys` remain consistent and unique, React can handle the DOM effectively and efficiently.

> ### Using Keys

Reviewer: ```suggestion
### Using keys
```

Style guide now uses sentence case

> Keys are passed into the component or a DOM element as a prop. You should already be familiar with the syntax.

Reviewer: should prop be marked in code as `prop` ? I haven't reviewed other lessons before so not sure if you decided on a standard here. With my limited knowledge on React this seems to be a react specific term. Similar to `argument` or `parameter`

Author: It doesn't really matter here, we could go with either honestly🤷

~~~jsx  
<Component key={keyValue} />
//or
<div key={keyValue} />
~~~

> <span id="keys-from-data">Now that we know the syntax, the next questions is what should be used as a key? Ideally they should be some identifier that is unique to each item in the list. Most databases assign a unique id to each entry, so you shouldn't have to worry about assigning and id yourself. If you are defining data yourself, it is good practice to assign a unique `id` to each item. You may use the [uuid](https://www.npmjs.com/package/uuid) package to generate a unique id. Let's look at an example:</span>

Reviewer: ```suggestion
<span id="keys-from-data">Now that we know the syntax, the next question is: what should be used as a key? Ideally they should be some identifier that is unique to each item in the list. Most databases assign a unique id to each entry, so you shouldn't have to worry about assigning and id yourself. If you are defining data yourself, it is good practice to assign a unique `id` to each item. You may use the [uuid](https://www.npmjs.com/package/uuid) package to generate a unique id. Let's look at an example:</span>
```

~~~jsx
const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

function MonthList(){
  return (
    <ul>
      {/* here we are using index as key */}
      {months.map((month, index) => (<li key={index}>{month}</li>))}
    </ul>
  )
}
~~~

<span id="anti-pattern">Keys are straightforward to use, though there is an anti-pattern you should be aware of. Keys should never be generated on the fly. Using `key={Math.random()}` or `key={uuid()}` _while_ rendering the list defeats the purpose of the list, as now a new `key` will ge created for every render of the list. As shown in the above example, `key` should be inferred from the data itself.</span>

~~~jsx
const todos = [
  { task: "mow the yard", id: uuid() }, 
  { task: "Work on Odin Projects", id: uuid() },
  { task: "feed the cat", id: uuid() },
];

function TodoList() {
  return (
    <>
      <ul>
        {todos.map((todo) => (
          // DON'T do the following i.e. generating keys during render    
          <li key={uuid()}>{todo.task}</li>
        ))}
      </ul>
    </>
  ) 
}

> ~~~

Reviewer 2: I think a common mistake as well, is having this `todos` array defined *inside* the component which re-creates the array on each re-render (or something similar), which can be bad as well. Maybe we need to also warn about stuff that gets re-created on each re-render, what do you think?

### Conclusion

> Don't fret if some of the terms covered in the lesson don't make sense yet. What's crucial right now is how to use keys effectively in React. As mentioned earlier, the more you learn about React, that's when you will realize why keys are so important. Furthermore, using keys is not limited to rendering lists. You might encounter use-cases where keys are needed, we'll leave that for you to discover.

Reviewer: ```suggestion
Don't fret if some of the terms covered in the lesson don't make sense yet. What's crucial right now is knowing how to use keys effectively in React. As mentioned earlier, the more you learn about React, that's when you will realize why keys are so important. Furthermore, using keys is not limited to rendering lists. You might encounter use-cases where keys are needed, we'll leave that for you to discover.
```

or alternatively something like:
> getting familiar with using keys effectively 

or:
> What's crucial right now is using keys effectively...

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Read this [section on keys in the React beta docs](https://beta.reactjs.org/learn/rendering-lists#keeping-list-items-in-order-with-key).

1. Watch this short video demonstrating [index as key being an anti-pattern](https://youtu.be/xlPxnc5uUPQ).

</div>

> ### Knowledge Check

Reviewer: ```suggestion
### Knowledge check
```
style guide

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

- [Why does React need keys?](#why-does-react-need-keys)
- [How do you use keys?](#using-keys)
- [Where should the key value ideally come from?](#keys-from-data)
- [When can we use array index as the key value?](#index-as-key)
- [What is an anti-pattern when using keys?](#anti-pattern)

> ### Additional Resources

Reviewer: ```suggestion
### Additional resources
```
style guide

### Introduction

> In this lesson we will cover keys in React. Keys are special props for our components and we'll learn why they are used.

Reviewer 3: Could possibly expand a little on the "why" without giving away the full story. 

```suggestion
We've already covered rendering lists, but we haven't covered an important part of that process: keys. Keys are special props in React that are incredibly useful when rendering items from an array, especially when it comes to re-rendering.
``

Or something like that maybe. We can entirely omit the first sentence ("In this lesson we will cover keys in React."), then just go into a quick tidbit about keys.

### Lesson overview

By the end of the lesson you should be able to answer the following:

- What are keys and why does React need them?
- What are good and bad examples of keys?

> ### Why does React need keys?

Reviewer 3: May not needed since we're linking to the React docs, but a "What are keys" section just to briefly explain them (plus an example that could help users understand them, like "Think of keys as this thing that may be familiar to you....")

In the previous lesson on rendering lists we used the `.map()` method to iterate over an array of data and return a list of elements. Now imagine, if any of the item in the list were to change, how would React know which item to update?

If the list were to change, one of two things _should_ happen:
1. we completely re-render the entire list, or:
2. we hunt down that specific items that were changed and only re-render those.

Assuming we want to hunt down that one specific item that was changed and NOT re-render the entire list. We need something to track that specific item. We can track down a specific item by using a `key`.

When the list is updated for whatever reason, (either from a server or a user interaction), React matches the `keys` of each of the previous list to the updated list. If there were any changes, React will only update the items that have changed.

As long as `keys` remain consistent and unique, React can handle the DOM effectively and efficiently.

### Using keys

Keys are passed into the component or a DOM element as a prop. You should already be familiar with the syntax.

~~~jsx  
<Component key={keyValue} />
//or
<div key={keyValue} />
~~~

> <span id="keys-from-data">Now that we know the syntax, the next question is: what should be used as a key? Ideally they should be some identifier that is unique to each item in the list. Most databases assign a unique id to each entry, so you shouldn't have to worry about assigning and id yourself. If you are defining data yourself, it is good practice to assign a unique `id` to each item. You may use the [uuid](https://www.npmjs.com/package/uuid) package to generate a unique id. Let's look at an example:</span>

Reviewer 3: Personally I feel like we could have used a non-3rd party package to show how to pass in keys, at least at first. For the purposes of the example it probably makes sense, but we could have as easily hardcoded unique `id` values here, then gone into using something like UUID and not generating keys on the fly further down.

~~~jsx
// a list of todos, each todo object has a task and an id
const todos = [
  { task: "mow the yard", id: uuid() }, 
  { task: "Work on Odin Projects", id: uuid() },
  { task: "feed the cat", id: uuid() },
];

function TodoList() {
  return (
    <ul>
      {todos.map((todo) => (
        // here we are using the already generated id as the key.
        <li key={todo.id}>{todo.task}</li>
      ))}
    </ul>
  ) 
}
~~~

> <span id="index-as-key">Additionally, if the list is static and will remain unchanged throughout the application's life, we can use the array index as a key. Though you should be careful when working with `index` as key.</span>

Reviewer 3: A quick sentence or two about why users should be careful when using index could be inserted here.

~~~jsx
const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];

function MonthList(){
  return (
    <ul>
      {/* here we are using index as key */}
      {months.map((month, index) => (<li key={index}>{month}</li>))}
    </ul>
  )
}
~~~

<span id="anti-pattern">Keys are straightforward to use, though there is an anti-pattern you should be aware of. Keys should never be generated on the fly. Using `key={Math.random()}` or `key={uuid()}` _while_ rendering the list defeats the purpose of the list, as now a new `key` will ge created for every render of the list. As shown in the above example, `key` should be inferred from the data itself.</span>

~~~jsx
const todos = [
  { task: "mow the yard", id: uuid() }, 
  { task: "Work on Odin Projects", id: uuid() },

>   { task: "feed the cat", id: uuid() },

Reviewer 3: Don't really need the `id` property in these objects for the purpose of this example

];

function TodoList() {
  return (
    <>
      <ul>
        {todos.map((todo) => (
          // DON'T do the following i.e. generating keys during render    
          <li key={uuid()}>{todo.task}</li>
        ))}
      </ul>
    </>
  ) 
}
~~~

### Conclusion

> Don't fret if some of the terms covered in the lesson don't make sense yet. What's crucial right now is knowing how to use keys effectively in React. As mentioned earlier, the more you learn about React, that's when you will realize why keys are so important. Furthermore, using keys is not limited to rendering lists. You might encounter use-cases where keys are needed, we'll leave that for you to discover.

Reviewer 3: Not sure if we really need this section... Not a huge thing, but in terms of the lesson it's not really a conclusion, and we probably don't need to tell users that lesson concepts may not make sense yet but will.
